File: Designer.Templates/Indicator/EmptyIndicator.fs
#nowarn "3391"

namespace StockSharp.Designer

open System
open Ecng.Common
open Ecng.Serialization

open StockSharp.Messages
open StockSharp.Algo
open StockSharp.Algo.Indicators

/// <summary>
/// Sample indicator demonstrating how to save and load parameters.
/// Changes the input price by +20% or -20%.
///
/// See more examples:
/// https://github.com/StockSharp/StockSharp/tree/master/Algo/Indicators
///
/// Documentation:
/// https://doc.stocksharp.com/topics/designer/strategies/using_code/fsharp/create_own_indicator.html
/// </summary>
type EmptyIndicator() as this =
    inherit BaseIndicator()

    // Internal fields
    let mutable changeValue = 20
    let mutable counter = 0
    let mutable isFormedValue = false
    // where the random decisions below come from - hand in a seeded source to repeat a run,
    // or a stub to state in a test what the values will be
    let mutable randomProvider: IRandomProvider = DefaultRandomProvider.Instance

    /// <summary>
    /// The percentage value (+/-) used to modify the input price.
    /// </summary>
    member this.Change
        with get () = changeValue
        and set value =
            changeValue <- value
            this.Reset()

    /// <summary>
    /// The source the random decisions are drawn from.
    /// </summary>
    member this.RandomProvider
        with get () = randomProvider
        and set value = randomProvider <- value

    /// <summary>
    /// Defines if the indicator has formed (became ready for trading).
    /// </summary>
    override this.CalcIsFormed() = isFormedValue

    /// <summary>
    /// Resets the indicator to its initial state.
    /// </summary>
    override this.Reset() =
        base.Reset()
        isFormedValue <- false
        counter <- 0

    /// <summary>
    /// The main logic to process input values.
    /// </summary>
    override this.OnProcess(input: IIndicatorValue) : IIndicatorValue =
        // every 10th call try to return an "empty" value
        if randomProvider.GetInt(0, 10) = 0 then
            // empty value still contains just time, no actual data
            DecimalIndicatorValue(this, input.Time)
        else
            // increment counter on each call
            counter <- counter + 1

            // after 5 inputs, indicator is considered formed
            if counter = 5 then
                isFormedValue <- true

            let mutable value = input.ToDecimal()

            // random change by a factor of +/- Change%
            let randomFactor = decimal (randomProvider.GetInt(-changeValue, changeValue)) / 100m
            value <- value + (value * randomFactor)

            // return final indicator value
            let result = DecimalIndicatorValue(this, value, input.Time)
            // randomly mark it as final or not
            result.IsFinal <- randomProvider.GetBool()
            result

    /// <summary>
    /// Load indicator settings from a given <see cref="SettingsStorage"/>.
    /// </summary>
    override this.Load(storage: SettingsStorage) =
        base.Load(storage)
        this.Change <- storage.GetValue<int>(nameof(this.Change))

    /// <summary>
    /// Save indicator settings to a given <see cref="SettingsStorage"/>.
    /// </summary>
    override this.Save(storage: SettingsStorage) =
        base.Save(storage)
        storage.SetValue(nameof(this.Change), this.Change)

    /// <summary>
    /// A string representation that includes the current <see cref="Change"/> value.
    /// </summary>
    override this.ToString() =
        sprintf "Change: %d" this.Change

File: Designer.Templates/Indicator/EmptyIndicator.cs
namespace StockSharp.Designer;

using System;
using System.ComponentModel;
using System.Linq;
using System.Collections.Generic;

using Ecng.Common;
using Ecng.Serialization;

using StockSharp.Messages;
using StockSharp.Algo;
using StockSharp.Algo.Indicators;

/// <summary>
/// Sample indicator demonstrating to save and load parameters.
/// 
/// Changes input price on +20% or -20%.
/// 
/// See more examples https://github.com/StockSharp/StockSharp/tree/master/Algo/Indicators
/// 
/// Doc https://doc.stocksharp.com/topics/designer/strategies/using_code/csharp/create_own_indicator.html
/// </summary>
public class EmptyIndicator : BaseIndicator
{
	private int _change = 20;

	public int Change
	{
		get => _change;
		set
		{
			_change = value;
			Reset();
		}
	}

	private int _counter;
	// formed indicator received all necessary inputs for be available for trading
	private bool _isFormed;

	protected override bool CalcIsFormed() => _isFormed;

	public override void Reset()
	{
		base.Reset();

		_isFormed = default;
		_counter = default;
	}

	protected override IIndicatorValue OnProcess(IIndicatorValue input)
	{
		// every 10th call try return empty value
		if (RandomGen.GetInt(0, 10) == 0)
			return new DecimalIndicatorValue(this, input.Time);

		if (_counter++ == 5)
		{
			// for example, our indicator needs 5 inputs for become formed
			_isFormed = true;
		}

		var value = input.ToDecimal();

		// random change on +20% or -20% current value

		value += value * RandomGen.GetInt(-Change, Change) / 100.0m;

		return new DecimalIndicatorValue(this, value, input.Time)
		{
			// final value means that this value for the specified input
			// is not changed anymore (for example, for candles that changes with last price)
			IsFinal = RandomGen.GetBool()
		};
	}

	// persist our properties to save for further the app restarts

	public override void Load(SettingsStorage storage)
	{
		base.Load(storage);
		Change = storage.GetValue<int>(nameof(Change));
	}

	public override void Save(SettingsStorage storage)
	{
		base.Save(storage);
		storage.SetValue(nameof(Change), Change);
	}

	public override string ToString() => $"Change: {Change}";
}